streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -93,43 +93,59 @@ public RecordCollectorImpl(final Producer<byte[], byte[]> producer, final String
+                    if (exception == null) {
+                        if (sendException != null) {
+                            return;
+                        }
+                        final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
+                        offsets.put(tp, metadata.offset());
+                    } else {
+                        if (sendException == null) {
+                            if (exception instanceof ProducerFencedException) {
+                                log.warn(LOG_MESSAGE, key, value, timestamp, topic, exception);
+                                sendException = new ProducerFencedException(
+                                    String.format(EXCEPTION_MESSAGE,
+                                                  logPrefix,
+                                                  "producer got fenced",
+                                                  key,
+                                                  value,
+                                                  timestamp,
+                                                  topic,
+                                                  exception.getMessage()));
+                            } else {
+                                String errorLogMessage = LOG_MESSAGE;
+                                String errorMessage = EXCEPTION_MESSAGE;
+                                if (exception instanceof RetriableException) {
+                                    errorLogMessage += PARAMETER_HINT;
+                                    errorMessage += PARAMETER_HINT;
                                 }
+                                log.error(errorLogMessage, key, value, timestamp, topic, exception);
+                                sendException = new StreamsException(
+                                    String.format(
+                                        EXCEPTION_MESSAGE,
+                                        errorMessage,
+                                        "an error caught",
+                                        key,
+                                        value,
+                                        timestamp,
+                                        topic,
+                                        exception), exception);
                             }
                         }
                     }
-                });
-                return;
-            } catch (final TimeoutException e) {
-                if (attempt == MAX_SEND_ATTEMPTS) {
-                    throw new StreamsException(String.format("%sFailed to send record to topic %s due to timeout after %d attempts", logPrefix, topic, attempt));
                 }
-                log.warn("Timeout exception caught when sending record to topic {}; retrying with {} attempt", topic, attempt);
-                Utils.sleep(SEND_RETRY_BACKOFF);
-            }
+            });
+        } catch (final TimeoutException e) {
```

Comment:
Although I agree that we should get rid of the retry logic but only rely on producer internals for retries, one caveat of this change is that if we did hit the issue that KIP-91 tries to solve, Streams will be less resilient as compared to today (previously we can tolerate up to 3X downtime while now it is X downtime) since increasing `max.block` will not help here.

Reply:
Would it not help if `max.block.ms` was set to a reasonably large value?

Reply:
@guozhangwang can you explain how this affects our ability to withstand downtime? You mentioned 3X vs. X downtime, what currently determines the value of X?

Reply:
Actually it's not exactly 3X v.s. X. And here is the difference:

Assuming the broker is down, then without this PR the producer would first use `request.timeout` to throw the exception for records in its accumulated queue, and then gets caught here and retry sending, and upon retries it will wait up to `max.block.ms` since queue is full and then throw the TimeoutException again, up to three times. So the total time it can endure broker to be down is 

`request.timeout + 3 * max.block.ms`

And without this PR it would be `request.timeout`.

Note that the issue itself will only happen if we do not yet know the destination leader of the partition when broker is down, so its likelihood-to-hit is not like 100%.

Reply:
I though the call to `send()` will never throw `TimeoutException` after `request.timeout` passed. The `TimeoutException` will be provided in the callback handler and streams will throw it later on. And this behavior does not change after this PR. One will always need to increase producer config `retries` to be more resilient to this scenario.

Please correct me if I am wrong.

Reply:
The `TimeoutException` that we catch here, should only originate from the scenario for a full producer queue and `max.block.ms` passed.

Reply:
@mjsax That is right, the `TimeoutException` from `Sender#failBatch()` is returned in the callback's exception, which will only be thrown in the next call. And `retries` will not help here. So it is really `max.block.ms` v.s. `3 * max.block.ms`.

Currently this config value's default is 60 secs and Streams does not override it. So the effect is that if we do hit the issue that KIP-91's solving, it is a resilience of 60 seconds v.s. 180 seconds.